### PR TITLE
Improve performance of build APK (~50%)  by running gen_snapshot concurrently

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Support concurrent build via flutter assemble. Compared to current master, a release build of an APK for target platforms android-arm, android-arm64, and android-x64 takes ~60 seconds, down from  ~140 seconds tested on my local macOS laptop.

Adds TargetPlatform.android enum. Longer-term, usage of TargetPlatform should be updated to remove arch specific flavors, these can be expressed with AndroidArch

Updated logs: https://gist.github.com/jonahwilliams/24aa5cf2544fd46e36a9b69a78d26ff4

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/file_hash_store.dart [hunk 1]
```diff
@@ -132,6 +132,9 @@ class FileHashStore {
     }
     final List<FileHash> fileHashes = <FileHash>[];
     for (MapEntry<String, String> entry in currentHashes.entries) {
+      previousHashes[entry.key] = entry.value;
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/file_hash_store.dart]: I don't need this change any more

File: packages/flutter_tools/gradle/flutter.gradle [hunk 2]
```diff
@@ -544,38 +544,36 @@ class FlutterPlugin implements Plugin<Project> {
...
-                    fileSystemRoots fileSystemRootsValue
-                    fileSystemScheme fileSystemSchemeValue
-                    trackWidgetCreation trackWidgetCreationValue
-                    targetPlatform targetArch
-                    sourceDir project.file(project.flutter.source)
-                    intermediateDir project.file("${project.buildDir}/${AndroidProject.FD_INTERMEDIATES}/flutter/${variant.name}/${targetArch}")
-                    extraFrontEndOptions extraFrontEndOptionsValue
-                    extraGenSnapshotOptions extraGenSnapshotOptionsValue
-                }
+            String taskName = toCammelCase(["compile", FLUTTER_BUILD_PREFIX, variant.name])
```
Comment (jonahwilliams) [inline packages/flutter_tools/gradle/flutter.gradle]: Instead of creating a task per ABI, I create a single task which builds all of the requested ABIs. This allows assemble to run concurrent gen_snapshots

File: packages/flutter_tools/gradle/flutter.gradle [hunk 3]
```diff
@@ -729,23 +733,15 @@ abstract class BaseFlutterTask extends DefaultTask {
             args "--depfile", "${intermediateDir}/flutter_build.d"
             args "--output", "${intermediateDir}"
             args "-dTargetFile=${targetPath}"
-            args "-dTargetPlatform=${targetPlatform}"
+            args "-dTargetPlatform=TargetPlatform.android"
```
Comment (zanderso) [inline packages/flutter_tools/gradle/flutter.gradle]: If the target platform is extracted from the target name, then why is this needed?

Comment (jonahwilliams) [reply packages/flutter_tools/gradle/flutter.gradle]: KernelSnapshot requires a target platform per your fuchsia changes. We could loosen that restriction a bit by making a missing platform default to non-fuchsia

Comment (jonahwilliams) [reply packages/flutter_tools/gradle/flutter.gradle]: We'd also still need it if we wanted to do platform specific assets at some point

File: packages/flutter_tools/gradle/flutter.gradle [hunk 4]
```diff
@@ -713,6 +706,17 @@ abstract class BaseFlutterTask extends DefaultTask {
...
 
+        // Compute the rule name for flutter assemble. To speed up builds that contain
+        // multiple ABIs, the target name is used to communicate which ones are required
+        // rather than the TargetPlatform. This allows multiple builds to share the same
+        // cache.
+        String[] ruleNames;
+        if (buildMode == "debug") {
+          ruleNames = ["debug_android_application"]
+        } else {
+          ruleNames = targetPlatformValues.collect({ targetPlatform ->
```
Comment (blasten) [inline packages/flutter_tools/gradle/flutter.gradle]: nit:
```.gradle
ruleNames = targetPlatformValues.collect { "android_aot_bundle_${buildMode}_$it" }
```

File: packages/flutter_tools/gradle/flutter.gradle [hunk 5]
```diff
@@ -768,7 +764,10 @@ class FlutterTask extends BaseFlutterTask {
             from "${intermediateDir}"
 
             if (buildMode == 'release' || buildMode == 'profile') {
-                include "app.so"
+                targetPlatformValues.each {
+                  String abi =  PLATFORM_ARCH_MAP[targetArch]
+                  include "${abi}/app.so"
```
Comment (blasten) [inline packages/flutter_tools/gradle/flutter.gradle]: nit:
```.gradle
include "${PLATFORM_ARCH_MAP[targetArch]}/app.so"
```

File: packages/flutter_tools/lib/src/build_info.dart [hunk 6]
```diff
@@ -306,6 +307,7 @@ String getNameForHostPlatform(HostPlatform platform) {
 }
 
 enum TargetPlatform {
+  android,
```
Comment (blasten) [inline packages/flutter_tools/lib/src/build_info.dart]: maybe add comment about what the difference is between  `android`, `android_arm` and so on.

Comment (blasten) [reply packages/flutter_tools/lib/src/build_info.dart]: we should probably just have `android` right?

File: packages/flutter_tools/lib/src/build_system/targets/android.dart [hunk 7]
```diff
@@ -122,3 +123,150 @@ class ReleaseAndroidApplication extends CopyFlutterAotBundle {
...
+  final TargetPlatform targetPlatform;
+
+  /// The selected build mode.
+  ///
+  /// This is restricted to [BuildMode.profile] or [BuildMode.relese].
+  final BuildMode buildMode;
+
+  @override
+  List<Source> get inputs => <Source>[
+    const Source.pattern('{FLUTTER_ROOT}/packages/flutter_tools/lib/src/build_system/targets/dart.dart'),
```
Comment (blasten) [inline packages/flutter_tools/lib/src/build_system/targets/android.dart]: q: why is this one part of the inputs?

File: packages/flutter_tools/gradle/flutter.gradle (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/gradle/flutter.gradle]: Done

File: packages/flutter_tools/gradle/flutter.gradle (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/gradle/flutter.gradle]: Done

Comment (jonahwilliams) [reply packages/flutter_tools/gradle/flutter.gradle]: Done

File: packages/flutter_tools/lib/src/build_info.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_info.dart]: Added a comment on the usage of android vs android w/ arch. Refactoring this is similar to the iOS refactoring

File: packages/flutter_tools/lib/src/build_system/targets/android.dart (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_system/targets/android.dart]: Sort of a work around to allow running from source to handle source file changes to the tool. It doesn't add much/any overhead to the build so it is okay to leave in.

File: dev/devicelab/bin/tasks/module_test.dart [hunk 8]
```diff
@@ -216,12 +216,12 @@ Future<void> main() async {
       }
 
       final String analyticsOutput = analyticsOutputFile.readAsStringSync();
-      if (!analyticsOutput.contains('cd24: android-arm64')
+      if (!analyticsOutput.contains('cd24: TargetPlatform.android')
```
Comment (zanderso) [inline dev/devicelab/bin/tasks/module_test.dart]: Is the real target architecture communicated in some other way to analytics?

**Comment by blasten** [comment]
The post update APK size is 43.5MB whereas it used to be 53.5MB. Do you know why it decreased in size?

File: dev/devicelab/bin/tasks/module_test.dart (hunk 8 above)
Comment (zanderso) [reply dev/devicelab/bin/tasks/module_test.dart]: If not, we should figure out a way to retain that so we can understand our analytics across this change.

**Comment by jonahwilliams** [comment]
I get 53.6MB building flutter gallery locally. Which app were you building?

**Comment by jonahwilliams** [comment]
Followed up offline: the logs were from a version that still had a bug in the app.so include logic. The latest commits build correctly

File: dev/devicelab/bin/tasks/module_test.dart (hunk 8 above)
Comment (jonahwilliams) [reply dev/devicelab/bin/tasks/module_test.dart]: It should be  part of the build apk/appbundle/aar commands:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/build_apk.dart#L58

@blasten can you provide some additional context on why we send this second analytics call? Its for the add2app tracking, right?

Comment (blasten) [reply dev/devicelab/bin/tasks/module_test.dart]: I don't think this is critical, but cc'ing @xster to confirm.

File: packages/flutter_tools/lib/src/build_info.dart (hunk 6 above)
Comment (blasten) [reply packages/flutter_tools/lib/src/build_info.dart]: sg

**Review by blasten — APPROVED** [review]

File: dev/devicelab/bin/tasks/module_test.dart (hunk 8 above)
Comment (xster) [reply dev/devicelab/bin/tasks/module_test.dart]: ultranit: not having the enum 'outer class' name in there would be nice as a bonus. 

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/build_apk.dart#L58 that you're linking to is orthogonal to this right? This is to detect top level calls to `flutter build bundles` from add-to-app scripts

**Comment by jonahwilliams** [comment]
YOLO

File: dev/devicelab/bin/tasks/module_test.dart (hunk 8 above)
Comment (xster) [reply dev/devicelab/bin/tasks/module_test.dart]: the dashboard to monitor is https://dashboards.corp.google.com/_8211ecd6_e372_4f1c_b4f8_a1f4f0411822#xhaf6f3g0

**Comment by dnfield** [comment]
This introduced https://github.com/flutter/flutter/issues/46163
